lib/Service/StackService.php

```diff
@@ -360,4 +376,103 @@ public function reorder($id, $order) {
 
 		return $result;
 	}
+
+	/**
+	 * @param $id
+	 * @param $boardId
+	 * @return Stack
+	 * @throws StatusException
+	 * @throws BadRequestException
+	 */
+	public function clone($id, $boardId) {
+		if (is_numeric($id) === false) {
+			throw new BadRequestException('stack id must be a number');
+		}
+		if (is_numeric($boardId) === false) {
+			throw new BadRequestException('board id must be a number');
+		}
+
+		$this->permissionService->checkPermission(null, $boardId, Acl::PERMISSION_MANAGE);
+		if ($this->boardService->isArchived(null, $boardId)) {
+			throw new StatusException('Operation not allowed. This board is archived.');
+		}
+
+		$stack = $this->stackMapper->find($id);
+		$board = $this->boardMapper->find($boardId);
+
+
+		$newStack = new Stack();
+		$newStack->setTitle($stack->getTitle() . ' (' . $this->l10n->t('copy') . ')');
+		$newStack->setBoardId($boardId);
+		$newStack->setOrder(999);
+		$newStack = $this->stackMapper->insert($newStack);
+
+		$this->activityManager->triggerEvent(
+			ActivityManager::DECK_OBJECT_BOARD, $newStack, ActivityManager::SUBJECT_STACK_CREATE
+		);
+		$this->changeHelper->boardChanged($boardId);
+
+		$this->eventDispatcher->dispatch(
+			'\OCA\Deck\Stack::onCreate',
+			new GenericEvent(null, ['id' => $newStack->getId(), 'stack' => $newStack])
+		);
+
+		$cards = $this->cardMapper->findAll($id);
+		$newCardArray = [];
+		foreach ($cards as $card) {
+
+			$newCard = new Card();
+			$newCard->setTitle($card->getTitle());
+			$newCard->setStackId($newStack->getId());
+			$newCard->setType($card->getType());
+			$newCard->setOrder($card->getOrder());
+			$newCard->setOwner($this->userId);
+			$newCard->setDescription($card->getDescription());
+			$newCard->setDuedate($card->getDuedate());
+
+			$newCard = $this->cardMapper->insert($newCard);
+
+			$this->activityManager->triggerEvent(ActivityManager::DECK_OBJECT_CARD, $newCard, ActivityManager::SUBJECT_CARD_CREATE);
```

Comment:
I'm not entirely sure if we should have one activity entry for each of the cards or one activity entry that the stack has been cloned, but we can improve that in a later step.